src/file_tree.ml

```diff
@@ -24,6 +25,47 @@ module Dune_file = struct
       | Ocaml_script of Path.t
   end
 
+  module Subdirs_stanza = struct
+    type t =
+      | Ignore_subdirs of Predicate_lang.t
+      | Subdirs of Predicate_lang.t
+
+    let decode : t Dune_lang.Decoder.t =
+      let ignore_subdirs =
+        let open Dune_lang.Decoder in
+        plain_string (fun ~loc dn ->
+          if Filename.dirname dn <> Filename.current_dir_name ||
+             match dn with
+             | "" | "." | ".." -> true
+             | _ -> false
+          then
+            of_sexp_errorf loc "Invalid sub-directory name %S" dn
+          else
+            dn)
+        |> list
+        >>| (fun l ->
+          Ignore_subdirs (Predicate_lang.of_string_set (String.Set.of_list l)))
+      in
+      let open Dune_lang.Decoder in
+      let subdirs =
+        Syntax.since Stanza.syntax (1, 6) >>>
+        Predicate_lang.decode >>| fun plang ->
+        Subdirs plang
+      in
+      sum [ "ignored_subdirs", ignore_subdirs
+          ; "subdirs", subdirs
+          ]
+
+    let reduce = function
+      | [] -> standard_dirs
+      | elems ->
+        let ignored, subdirs = List.partition_map elems ~f:(function
+          | Ignore_subdirs s -> Either.Left s
+          | Subdirs d -> Either.Right d)
+        in
+        Predicate_lang.(diff (union subdirs) (union ignored))
```

Comment:
It feels like if we only have an `ignored_subdirs` stanza, then this is `<empty> && (not <ignored>)` and so matches nothing. Is that effectively the case? and if not why?

Reply:
Ah yeah good call. Fixed.